src/sink/gcs.rs

```diff
@@ -0,0 +1,345 @@
+                    storage::download_object(&remote, &bucket_name, &object_name).await?
+                }
+                StorageCommand::Unknown => {
+                    warn!(
+                        "Unknown Google Cloud Storage command: `{}` attempted",
+                        command.to_string()
+                    );
+
+                    return Err(format!(
+                        "Unknown Google Cloud Storage command: `{}` attempted",
+                        command.to_string()
+                    )
+                    .into());
+                }
+            };
+            let response: Value = json!({
+                "cmd": command_str,
+                "data": response
+            })
+            .into();
+            if self.is_linked {
+                if let Some(reply_channel) = &self.reply_channel {
+                    reply_channel
+                        .send(sink::Reply::Response(
+                            OUT,
+                            Event {
+                                id: EventId::new(1, 2, 3),
+                                data: LineValue::new(vec![], |_| ValueAndMeta::from(response)),
+                                ingest_ns: nanotime(),
+                                origin_uri: Some(EventOriginUri {
+                                    uid: 0,
+                                    scheme: "gRPC".into(),
+                                    host: "".into(),
+                                    port: None,
+                                    path: vec![],
+                                }),
+                                kind: None,
+                                is_batch: false,
+                                cb: CbAction::None,
+                                op_meta: OpMeta::default(),
+                                transactional: false,
+                            },
+                        ))
+                        .await?;
+                }
+            }
+        }
+
+        self.is_down = false;
+        return Ok(Some(vec![qos::ack(&mut event)]));
+    }
+
+    fn default_codec(&self) -> &str {
+        "json"
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    async fn init(
+        &mut self,
+        _sink_uid: u64,
+        _sink_url: &TremorUrl,
+        _codec: &dyn Codec,
+        _codec_map: &HashMap<String, Box<dyn Codec>>,
+        _processors: Processors<'_>,
+        is_linked: bool,
+        reply_channel: Sender<sink::Reply>,
+    ) -> Result<()> {
+        self.reply_channel = Some(reply_channel);
+        self.is_linked = is_linked;
+        Ok(())
+    }
+    async fn on_signal(&mut self, signal: Event) -> ResultVec {
+        if self.is_down && self.qos_facility.probe(signal.ingest_ns) {
+            self.is_down = false;
+            // This means the port is connectable
+            info!("Google Cloud Storage -  sink remote endpoint - recovered and contactable");
+            self.is_down = false;
```

Comment:
We won't need to set this twice :)

Reply:
This is boilerplate as the failover/detector is yet to be tested